### PR TITLE
feat: add basic connectivity check

### PR DESCRIPTION
**Pull request by cesfahani**
Implement a simple connectivity check in a new `gix-fsck` crate, and add this to `gix` via a new `fsck` subcommand. Currently this is functionally equivalent to:
`git rev-list --objects --quiet --missing=print`

### Tasks

*Please note that these tasks are notes of Byron, feel free to make it yours though.*

* [x] update `crate-status.md` with information about what's done and what is not, especially in comparison to what `git fsck` really does.
* [x] update `README.md` with new crate

### Tasks for Byron

* [x] refactor
* [x] flatten `fsck` command to be more similar to `fsck` (in lack of a vision what the other subcommands would be)
* [x] ~~try to improve gix-testtools by configuring the preamble in the fixture script used here~~ - does not work, it's very particular about where it wants to find that settings, needs it to be in globals. Fair enough.

### CODE REVIEW

File: gix-fsck/src/lib.rs
Comment (Byron) [inline gix-fsck/src/lib.rs:45]: Now that this functionality is in a crate, it needs to be fully tested.

However, if it's too early for that, it's fine to move it all into `gitoxide-core` which implies it's pre-crate experimentation stage that will only be tested by hand, by running it on repositories.

**Review by Byron — CHANGES_REQUESTED** [review]
Thanks a lot for contributing, great work!

I took a first look and the most impactful remark I have is that either the new crate needs tests, or the code should move to `gitoxide-core` for experimentation.

Thanks and cheers

**Comment by cesfahani** [comment]
I seem to have slightly botched this PR... I spent some time iterating on the formatting CI checks and fixed them, but then it looks like I failed to push the fixes before submitting the draft PR :face_exhaling:. This revision of the PR also includes the incorrect committer email address (my work one). I'll get this fixed.

Your comments are all still perfectly valid as they don't relate to the formatting checks - replies incoming...

File: gix-fsck/LICENSE-APACHE
Comment (cesfahani) [reply gix-fsck/LICENSE-APACHE:1]: Nope - that was entirely a mistake. Will fix (assuming I end up keeping the new crate at all).

File: gix-fsck/LICENSE-MIT
Comment (cesfahani) [reply gix-fsck/LICENSE-MIT:1]: Same as above - my bad.

File: gix-fsck/src/lib.rs
Comment (cesfahani) [reply gix-fsck/src/lib.rs:45]: I did try this, but found that because I end up using a `HashSet` of all objects the connectivity check has already encountered, I'll end up needing to create an `ObjectId` out of the `oid` for *every* object anyways. After coming to that conclusion, it seemed to me it'd be more efficient to just convert to an `ObjectId` up-front and use that throughout the connectivity check routines.

I'm probably missing something here... If so, please let me know.

File: gix/Cargo.toml
Comment (cesfahani) [reply gix/Cargo.toml:255]: Will do, thanks!

File: gix-fsck/src/lib.rs
Comment (cesfahani) [reply gix-fsck/src/lib.rs:45]: I've been working on adding some tests for this crate and plan on getting those into this PR before moving it out of draft. It's taking me a bit longer than I expected, mainly because I'm trying to get the test fixtures to produce deterministic repositories (at least with the assumption that the host machine has a somewhat recent `git` version). I see you've done some similar things around the existing tests.

I have no objections to ditching this new crate and moving it to `gitoxide-core`, but either way, I'd really like to have tests that make sure I'm not breaking anything as I progress.

Comment (Byron) [reply gix-fsck/src/lib.rs:45]: Yes, you are right. I also end up using `&ObjectId` when hashsets are involved, I was a bit quick there with the armchair review.

Comment (Byron) [reply gix-fsck/src/lib.rs:45]: Yes, tests are definitely a good idea even though they aren't a requirement while it's experimental, particularly if it would move to `gitoxide-core`. However, I think once you have added tests, the code should stay in the crate which, as a container for sanity checks of the structure itself, does make sense.

File: src/plumbing/main.rs
Comment (cesfahani) [reply src/plumbing/main.rs:1075]: I think I may have misunderstood what you meant here.

Per the comment above, I removed all references of `gix-fsck` from the `gix` crate. But I still have a small wrapper around `gix-fsck` in the `gitoxide-core` crate. Did you mean I should move that from `gitoxide-core` to here? If so, wouldn't I need to add `gix-fsck` to the `[dependency]` list in the top-level `Cargo.toml`?

I'm happy to rearrange things, just wanted to clarify.

File: gix-fsck/src/lib.rs
Comment (cesfahani) [reply gix-fsck/src/lib.rs:45]: I've added a few tests in my latest update that exercise full, blobless, and treeless clones. There could definitely be more added, but wanted to get something in front of you sooner rather than later so you could see if I've done anything wrong with the test setup (particularly with the test fixture generation - let me know if I went off into the weeds there).

File: gix-fsck/tests/fsck/connectivity/mod.rs
Comment (Byron) [inline gix-fsck/tests/fsck/connectivity/mod.rs:40]: I think this would be called `hex_to_ids` to stay in line with `hex_to_id` which is used everywhere.
In theory, this utility could be provided by `gix-testtools` as well.

File: src/plumbing/main.rs
Comment (Byron) [reply src/plumbing/main.rs:1075]: I think this is good now, thank you :).

**Review by Byron — COMMENTED** [review]
Thanks again for keeping at it!

Besides leaving a few comments which I hope are descriptive, I will provide you with a new trait that helps you to avoid using `gix-odb` as dependency.

And even though this PR is still some steps away from being mergable, it's getting there and I think the learning experience will be valuable to you. If that's ever not the case or you feel exhausted, please do let me know as I can also push it to completion myself then, no problem.

**Comment by Byron** [comment]
Once https://github.com/Byron/gitoxide/pull/1088 is merged, you could rebase remove the `gix-odb` dependency, please feel free to subscribe for notifications.

File: gix-fsck/tests/Cargo.toml
Comment (cesfahani) [reply gix-fsck/tests/Cargo.toml]: Done, and thank you! I was a bit confused because by attempt at partial cloning using the existing test tooling was causing my tests to fail, but that ended up being due to `uploadpack.allowFilter`not being set in persistent git config. See below for details on that...

File: gix-fsck/tests/fixtures/make_test_repos.sh
Comment (cesfahani) [reply gix-fsck/tests/fixtures/make_test_repos.sh]: Ack!

File: gix-fsck/tests/fixtures/make_test_repos.sh
Comment (cesfahani) [reply gix-fsck/tests/fixtures/make_test_repos.sh]: Ack!

File: gix-fsck/tests/fixtures/make_test_repos.sh
Comment (cesfahani) [reply gix-fsck/tests/fixtures/make_test_repos.sh]: Ack!

File: gix-fsck/tests/fixtures/make_test_repos.sh
Comment (cesfahani) [reply gix-fsck/tests/fixtures/make_test_repos.sh]: Nope - no problem for my test at all. I've remove all that cruft and have left only the one piece I think I still need...

I'm finding that without the line that sets `GIT_CONFIG_GLOBAL` to a non-empty value, and without the line that sets `uploadpack.allowFilter = true` in that global config, I end up being unable to do partial clones in the test environment. The clones "work", but they spew out a warning that the server doesn't support partial clones, and they fall-back to doing a full clone instead. Luckily that causes my unit tests to fail, so it's easy to spot. 

I haven't dug too much into how the "smart" git transport works when cloning from a path on the local machine, but I expect it just forks/execs a separate `git` process to act as the server. It also seems to sanitize the environment before doing so, which makes sense, as that means it'll behave much closer to dealing with a real remote. It seems like it respects/preserves the `GIT_CONFIG_GLOBAL` variable, but ignores the `GIT_CONFIG_KEY_X` and `GIT_CONFIG_VALUE_X` variables. If you'd like me to dig in a bit and confirm where this happens in the `git` source, just say the word.

I've left in the line that sets & exports the `GIT_CONFIG_GLOBAL` variable and the line that sets `uploadpack.allowFilter` in that config, but please let me know if you have other ideas of how to handle this one. I did consider using lower-level `git` commands to produce test fixtures, but that seems like it'll be a pain later on when I'm trying to test a partial clone implementation...

File: gix-fsck/tests/fixtures/make_test_repos.sh
Comment (cesfahani) [reply gix-fsck/tests/fixtures/make_test_repos.sh]: Thanks! I added the archive to `.gitignore` in my latest changes.

File: gix-fsck/tests/fsck/connectivity/mod.rs
Comment (cesfahani) [reply gix-fsck/tests/fsck/connectivity/mod.rs]: This function doesn't just spit out a collection of `ObjectId`'s, but rather, it provides a map of `ObjectID`s to `Kind`'s. The function above this does exactly what you're implying, and it is named as you'd expect (`hex_to_ids`).

I understand that "objects" probably isn't the best naming here, as it is not outputting actual objects, but rather a representation of their (expected) metadata. Totally open to better naming!

**Comment by cesfahani** [comment]
> Once #1088 is merged, you could rebase remove the `gix-odb` dependency, please feel free to subscribe for notifications.

Done!

**Comment by Byron** [comment]
Thanks!

I should be able to finish the review today - it will be an active one so I will look at everything and make changes as needed. These I will push while rewriting the commit-queue, possibly multiple times - please refrain to make any further changes to avoid complications.

Even though the gist of changes will be described in their respective commits, I will provide a summary here before merging with everything I think could help further the feature.

File: gix-fsck/tests/fixtures/make_test_repos.sh
Comment (Byron) [reply gix-fsck/tests/fixtures/make_test_repos.sh]: I did also try to add this small preamble to `gix-testtools`, but like you mentioned already, it's indeed a settings that needs to be in the global configuration.

**Review by Byron — APPROVED** [review]
Thanks a lot! If you are interested, you can take a look at the added commits to participate in my refactoring-journey.

A few notes:

* when traversing trees, the `gix-traverse` crate can be used for that. However, in this case it seems easier to just implement traversal ourselves.
* Please avoid recursion, as doing so removes an attack vector and typically leads to a better time with the borrowchecker - win-win if you will :).

### Performance

On the `git` repository, `git fsck --connectivity-only`, which still performs more checking than is implemented here, will run in ~2.5s, while `gix fsck` takes ~6s.

```
❯ hyperfine -N -w 1 'git fsck --connectivity-only' '/Users/byron/dev/github.com/Byron/gitoxide/target/release/gix fsck'
Benchmark 1: git fsck --connectivity-only
  Time (mean ± σ):      2.461 s ±  0.022 s    [User: 2.333 s, System: 0.090 s]
  Range (min … max):    2.415 s …  2.491 s    10 runs

Benchmark 2: /Users/byron/dev/github.com/Byron/gitoxide/target/release/gix fsck
  Time (mean ± σ):      6.278 s ±  0.026 s    [User: 6.216 s, System: 0.018 s]
  Range (min … max):    6.253 s …  6.341 s    10 runs

Summary
  git fsck --connectivity-only ran
    2.55 ± 0.03 times faster than /Users/byron/dev/github.com/Byron/gitoxide/target/release/gix fsck
```


It was somewhat expected that this happens as `git` - somehow - manages to access all objects much more efficiently. `gix` spends the majority of its time decoding deltified objets, with the pack-cache for deltas getting just a few hits. This leads to a huge waste of time.

Screenshot 2023-11-10 at 09 37 09

Screenshot 2023-11-10 at 09 38 05

What I feel uncomfortable with is also that `fsck` is actually *a lot more* than what's currently implemented, as technically it's only a check for missing objects. Even though I tried to be upfront about this in the command's documentation, this in conjunction with the performance issue makes me want to spend more time on it.

Something I'd love to see is if there is something in `git fsck` that allows git to be so much faster or more efficient when accessing the ODB. Running `git fsck --connectivity-only -v` shows that commits seemingly have been ordered 'a little'. Maybe it's worth it to study their algorithm and try to mimic it, maybe to get access patterns that improve the object access performance which now is the major bottle-neck.

Please let me know what you think about this.

*A simple solution to one of these problems would be to rename `Connectivity` into something that shows what it's currently doing, and not what it aspires to be. The performance issue would remain though.*

**Comment by Byron** [comment]
### Correction

Performance isn't as simple apparently, as on the linux repository, I see 46s for `git` and only 22s for `gix` - this would give room for additional checks to make me think that `gix` isn't always slower.

```
❯ hyperfine -N -w 1 'git fsck --connectivity-only' '/Users/byron/dev/github.com/Byron/gitoxide/target/release/gix fsck'
Benchmark 1: git fsck --connectivity-only
  Time (mean ± σ):     46.783 s ±  1.158 s    [User: 45.172 s, System: 0.808 s]
  Range (min … max):   45.004 s … 48.705 s    10 runs

Benchmark 2: /Users/byron/dev/github.com/Byron/gitoxide/target/release/gix fsck
  Time (mean ± σ):     21.987 s ±  0.556 s    [User: 21.639 s, System: 0.190 s]
  Range (min … max):   21.242 s … 22.689 s    10 runs

Summary
  /Users/byron/dev/github.com/Byron/gitoxide/target/release/gix fsck ran
    2.13 ± 0.08 times faster than git fsck --connectivity-only
```

**Comment by Byron** [comment]
Another note: maybe it really does need a special algorithm. Maybe it's just unrelated though: https://github.com/praetorian-inc/noseyparker/blob/3b160d86eca484b36ff754fab7337368547af704/crates/input-enumerator/src/git_metadata_graph.rs#L313 , but I thought I'd share anyway in case it can inspire to find novel solutions to old problems.

**Pull request merged**
